src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
+        return {ParseStatus::OK, m_request, headerLength};
+    }
+    else if (m_request.method == HEADER_REQUEST_METHOD_POST) {
+        bool ok = false;
+        const int contentLength = m_request.headers[HEADER_CONTENT_LENGTH].toInt(&ok);
+        if (!ok || (contentLength < 0)) {
+            qWarning() << Q_FUNC_INFO << "bad request: content-length invalid";
+            return {ParseStatus::BadRequest, Request(), 0};
+        }
+        if (contentLength > MAX_REQUEST_SIZE) {
             qWarning() << Q_FUNC_INFO << "bad request: message too long";
-            return BadRequest;
+            return {ParseStatus::BadRequest, Request(), 0};
         }
 
-        QByteArray content = data.mid(headerEnd + EOH.length(), contentLength);
-        if (content.length() < contentLength) {
-            qDebug() << Q_FUNC_INFO << "incomplete request";
-            return IncompleteRequest;
-        }
+        if (contentLength > 0) {
+            const QByteArray httpBody = data.mid(headerLength, contentLength);
+            if (httpBody.length() < contentLength) {
+                qDebug() << Q_FUNC_INFO << "incomplete request";
+                return {ParseStatus::Incomplete, Request(), 0};
+            }
+            if (httpBody.length() != contentLength) {
+                qDebug() << Q_FUNC_INFO << "content length mismatch";
+                return {ParseStatus::BadRequest, Request(), 0};
+            }
 
-        if ((contentLength > 0) && !parseContent(content)) {
-            qWarning() << Q_FUNC_INFO << "message parsing error";
-            return BadRequest;
+            if (!parsePostMessage(httpBody)) {
+                qWarning() << Q_FUNC_INFO << "message body parsing error";
+                return {ParseStatus::BadRequest, Request(), 0};
+            }
         }
-    }
 
-    //  qDebug() << Q_FUNC_INFO;
-    //  qDebug() << "HTTP Request header:";
-    //  qDebug() << data.left(headerEnd) << "\n";
+        return {ParseStatus::OK, m_request, (headerLength + contentLength)};
+    }
 
-    request = m_request;
-    return NoError;
+    qWarning() << Q_FUNC_INFO << "unsupported request method: " << m_request.method;
+    return {ParseStatus::BadRequest, Request(), 0};  // TODO: SHOULD respond "501 Not Implemented"
 }
 
-bool RequestParser::parseStartingLine(const QString &line)
+bool RequestParser::parseStartLines(const QString &data)
 {
-    const QRegExp rx("^([A-Z]+)\\s+(\\S+)\\s+HTTP/\\d\\.\\d$");
-
-    if (rx.indexIn(line.trimmed()) >= 0) {
-        m_request.method = rx.cap(1);
-
-        QUrl url = QUrl::fromEncoded(rx.cap(2).toLatin1());
-        m_request.path = url.path(); // Path
-
-        // Parse GET parameters
-        QListIterator<QPair<QString, QString> > i(QUrlQuery(url).queryItems());
-        while (i.hasNext()) {
-            QPair<QString, QString> pair = i.next();
-            m_request.gets[pair.first] = pair.second;
+    // for simplicity, we don't handle malformed request which uses `LF` for newline
```

Comment:
request which uses `LF` for newline is bad request!

Reply:
AFAIK some program provides compatibility, but we don't, should I change the comment?

Reply:
It looks like we're sorry we don't support the malformed requests.

Reply:
I'll remove the "for simplicity"